cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/migrations/MigrateDatabase100To200.java

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright © 2016 IBM Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+package com.cloudant.sync.datastore.migrations;
+
+import com.cloudant.android.ContentValues;
+import com.cloudant.sync.datastore.DatastoreImpl;
+import com.cloudant.sync.sqlite.Cursor;
+import com.cloudant.sync.sqlite.SQLDatabase;
+import com.cloudant.sync.util.CouchUtils;
+import com.cloudant.sync.util.DatabaseUtils;
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * <p>
+ * Migration to repair datastores impacted by issues 326 and 329.
+ * </p>
+ * <p>
+ * Added migration on `Datastore` opening to repair datastores that have documents with
+ * multiple identical revision IDs (as caused by issue #329). The migration will delete duplicate
+ * revisions and correct the document tree. It will also re-evaluate the winning revision resolving
+ * issues caused by #326.
+ * </p>
+ * <p>
+ * It also deletes duplicate entries with the same sequence and attachment filename, which could
+ * have been force inserted into the attachments table during a multipart attachment pull during
+ * identical replications running in parallel.
+ * </p>
+ */
+public class MigrateDatabase100To200 implements Migration {
+
+    @Override
+    public void runMigration(SQLDatabase db) throws Exception {
+        Cursor c = null;
+        try {
+            // Find docs with duplicate revision ids and get the lowest sequence for that revision
+            List<DocRevSequence> lowestDuplicateRevs = new ArrayList<DocRevSequence>();
+            c = db.rawQuery("SELECT doc_id, revid, min(sequence) FROM revs GROUP BY doc_id, revid" +
+                    " HAVING COUNT(*) > 1;", null);
```

Comment:
I think on android `null` needs to be an empty array, it caught me out with the tests. Waiting on CI to confirm though.